File: composeApp/src/androidMain/kotlin/com/arcvgc/app/ui/contentlist/ContentListViewModel.kt
package com.arcvgc.app.ui.contentlist

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arcvgc.app.data.repository.AppConfigRepository
import com.arcvgc.app.data.repository.AppConfigRepositoryImpl
import com.arcvgc.app.data.repository.BattleRepositoryImpl
import com.arcvgc.app.data.repository.FavoritesRepository
import com.arcvgc.app.data.repository.FavoritesRepositoryImpl
import com.arcvgc.app.data.repository.FormatCatalogRepository
import com.arcvgc.app.data.repository.PokemonCatalogRepository
import com.arcvgc.app.data.repository.SettingsRepository
import com.arcvgc.app.domain.model.LookbackWindow
import com.arcvgc.app.domain.model.OrderBy
import com.arcvgc.app.ui.model.ContentListMode
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ContentListViewModel @Inject constructor(
    private val battleRepositoryImpl: BattleRepositoryImpl,
    private val favoritesRepositoryImpl: FavoritesRepositoryImpl,
    val settingsRepository: SettingsRepository,
    private val pokemonCatalogRepository: PokemonCatalogRepository,
    private val appConfigRepositoryImpl: AppConfigRepositoryImpl,
    val formatCatalogRepository: FormatCatalogRepository
) : ViewModel() {

    private var logic: ContentListLogic? = null

    val favoritesRepository: FavoritesRepository get() = favoritesRepositoryImpl
    val appConfigRepository: AppConfigRepository get() = appConfigRepositoryImpl

    private val _uiState = MutableStateFlow(ContentListUiState())
    val uiState: StateFlow<ContentListUiState> = _uiState.asStateFlow()

    private val _sortOrder = MutableStateFlow(OrderBy.Time)
    val sortOrder: StateFlow<OrderBy> = _sortOrder.asStateFlow()

    private val _selectedFormatId = MutableStateFlow(0)
    val selectedFormatId: StateFlow<Int> = _selectedFormatId.asStateFlow()

    private val _selectedLookback = MutableStateFlow(LookbackWindow.All)
    val selectedLookback: StateFlow<LookbackWindow> = _selectedLookback.asStateFlow()

    // Apply-once tick pattern for deep-linked lookback: each deep link that
    // carries a lookback bumps a tick on the composition side; this counter
    // tracks the most recent tick already applied so re-entering composition
    // (tab switches, recomposition) with the same tick doesn't clobber a
    // lookback the user has since changed in the UI.
    var lastAppliedLookbackTick: Int = 0

    private var initialized = false

    fun initialize(mode: ContentListMode, initialLookback: LookbackWindow = LookbackWindow.All) {
        if (initialized) return
        initialized = true
        val l = ContentListLogic(
            scope = viewModelScope,
            repository = battleRepositoryImpl.shared,
            favoritesRepository = favoritesRepositoryImpl.shared,
            appConfigRepository = appConfigRepositoryImpl.shared,
            mode = mode,
            pokemonCatalogItems = pokemonCatalogRepository.state.value.items,
            pokemonCatalogState = pokemonCatalogRepository.state,
            initialLookback = initialLookback,
            settingsRepository = settingsRepository.shared,
            isFormatHistoric = { id ->
                formatCatalogRepository.state.value.items.find { it.id == id }?.isHistoric == true
            }
        )
        logic = l

        viewModelScope.launch { l.uiState.collect { _uiState.value = it } }
        viewModelScope.launch { l.sortOrder.collect { _sortOrder.value = it } }
        viewModelScope.launch { l.selectedFormatId.collect { _selectedFormatId.value = it } }
        viewModelScope.launch { l.selectedLookback.collect { _selectedLookback.value = it } }
        viewModelScope.launch { l.searchQuery.collect { _searchQuery.value = it } }

        l.initialize()
    }

    fun loadContent() { logic?.loadContent() }
    fun refresh() { logic?.refresh() }
    fun paginate() { logic?.paginate() }
    fun selectFormat(formatId: Int) { logic?.selectFormat(formatId) }
    fun selectLookback(lookback: LookbackWindow) { logic?.selectLookback(lookback) }
    fun toggleSortOrder() { logic?.toggleSortOrder() }

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    fun setSearchQuery(query: String) { logic?.setSearchQuery(query) }
}

File: iosApp/iosApp/ContentListViewModel.swift
import Foundation
import Shared

enum FavoriteContentType {
    case battles, pokemon, players
}

enum ContentListMode {
    case home
    case favorites(contentType: FavoriteContentType)
    case search(params: SearchParams)
    case pokemon(id: Int32, name: String, imageUrl: String?, typeImageUrl1: String?, typeImageUrl2: String?, formatId: Int32? = nil)
    case player(id: Int32, name: String, formatId: Int32? = nil)
    case topPokemon(formatId: Int32? = nil)

    func toSharedMode() -> Shared.ContentListMode {
        switch self {
        case .home:
            return Shared.ContentListMode.Home.shared
        case .favorites(let contentType):
            let sharedType: Shared.FavoriteContentType = {
                switch contentType {
                case .battles: return .battles
                case .pokemon: return .pokemon
                case .players: return .players
                }
            }()
            return Shared.ContentListMode.Favorites(contentType: sharedType)
        case .search(let params):
            return Shared.ContentListMode.Search(params: params)
        case .pokemon(let id, let name, let imageUrl, let typeImageUrl1, let typeImageUrl2, let formatId):
            return Shared.ContentListMode.Pokemon(
                pokemonId: id,
                name: name,
                imageUrl: imageUrl,
                typeImageUrl1: typeImageUrl1,
                typeImageUrl2: typeImageUrl2,
                formatId: formatId.map { KotlinInt(int: $0) },
                lookback: nil
            )
        case .player(let id, let name, let formatId):
            return Shared.ContentListMode.Player(
                playerId: id,
                playerName: name,
                formatId: formatId.map { KotlinInt(int: $0) }
            )
        case .topPokemon(let formatId):
            return Shared.ContentListMode.TopPokemon(
                formatId: formatId.map { KotlinInt(int: $0) }
            )
        }
    }
}

@MainActor
final class ContentListViewModel: ObservableObject {
    @Published private(set) var state: ContentListUiState
    @Published private(set) var sortOrder: OrderBy
    @Published private(set) var selectedFormatId: Int32
    @Published private(set) var selectedLookback: LookbackWindow
    @Published private(set) var searchQuery: String
    @Published private(set) var mode: ContentListMode

    // Mirrors the Android/web `lastAppliedLookbackTick` pattern: each deep
    // link that carries a lookback bumps a tick in `ContentView`; this counter
    // tracks the most recent tick already applied so re-attaching the view
    // (tab re-render with the same tick) doesn't clobber a lookback the user
    // has since changed in the UI.
    var lastAppliedLookbackTick: Int32 = 0

    let formatItems: [FormatUiModel]

    private let logic: ContentListLogic
    private let logicScope: Kotlinx_coroutines_coreCoroutineScope
    private var observationTasks: [Task<Void, Never>] = []

    init(repository: BattleRepository, mode: ContentListMode = .home, favoritesStore: FavoritesStore? = nil, pokemonCatalogItems: [PokemonPickerUiModel] = [], appConfigStore: AppConfigStore? = nil, formatItems: [FormatUiModel] = [], settingsStore: SettingsStore? = nil, initialLookback: LookbackWindow = .all) {
        self.mode = mode
        self.formatItems = formatItems

        let scope = CoroutineScopeFactory.shared.createMainScope()
        self.logicScope = scope

        let sharedMode = mode.toSharedMode()
        let formatItemsCopy = formatItems
        self.logic = ContentListLogic(
            scope: scope,
            repository: repository,
            favoritesRepository: favoritesStore?.repo ?? FavoritesRepository(storage: FavoritesStorage()),
            appConfigRepository: appConfigStore?.repo ?? AppConfigRepository(
                apiService: ApiService(),
                storage: AppConfigStorage(),
                catalogCacheStorage: CatalogCacheStorage()
            ),
            mode: sharedMode,
            pokemonCatalogItems: pokemonCatalogItems,
            pokemonCatalogState: nil,
            initialTopPokemonFetchCount: 6,
            initialLookback: initialLookback,
            settingsRepository: settingsStore?.repo,
            isFormatHistoric: { id in
                let isHistoric = formatItemsCopy.first(where: { $0.id == id.int32Value })?.isHistoric ?? false
                return KotlinBoolean(bool: isHistoric)
            }
        )

        self.state = logic.uiState.value
        self.sortOrder = (logic.sortOrder.value as? OrderBy) ?? .time
        self.selectedFormatId = (logic.selectedFormatId.value as! KotlinInt).int32Value
        self.selectedLookback = (logic.selectedLookback.value as? LookbackWindow) ?? .all
        self.searchQuery = logic.searchQuery.value as String

        observationTasks.append(Task { [weak self] in
            for await uiState in logic.uiState {
                self?.state = uiState
            }
        })
        observationTasks.append(Task { [weak self] in
            for await order in logic.sortOrder {
                self?.sortOrder = (order as? OrderBy) ?? .time
            }
        })
        observationTasks.append(Task { [weak self] in
            for await id in logic.selectedFormatId {
                self?.selectedFormatId = (id as! KotlinInt).int32Value
            }
        })
        observationTasks.append(Task { [weak self] in
            for await lookback in logic.selectedLookback {
                self?.selectedLookback = (lookback as? LookbackWindow) ?? .all
            }
        })
        observationTasks.append(Task { [weak self] in
            for await query in logic.searchQuery {
                self?.searchQuery = query as String
            }
        })

        logic.initialize()
    }

    func loadContent() { logic.loadContent() }
    func refresh() { logic.refresh() }
    func paginate() { logic.paginate() }
    func selectFormat(_ formatId: Int32) { logic.selectFormat(formatId: formatId) }
    func selectLookback(_ lookback: LookbackWindow) { logic.selectLookback(lookback: lookback) }
    func toggleSortOrder() { logic.toggleSortOrder() }
    func setSearchQuery(_ query: String) { logic.setSearchQuery(query: query) }

    func updateSearchParams(_ params: SearchParams) {
        mode = .search(params: params)
        logic.updateSearchParams(params: params)
    }

    deinit {
        observationTasks.forEach { $0.cancel() }
        CoroutineScopeFactory.shared.cancel(scope: logicScope)
    }
}
